### PR TITLE
speedup code for computing RDMs

### PR DESCRIPTION
**Pull request by HeikoSchuett**
here specifically skip descriptor things for averaging datasets and save some computation for dataset handling in-between

### CODE REVIEW

**Comment by HeikoSchuett** [comment]
This is in response to #211

File: rsatoolbox/data/computations.py [hunk 1]
```diff
@@ -20,7 +21,7 @@ def average_dataset(dataset):
     return np.mean(dataset.measurements, axis=0)
 
 
-def average_dataset_by(dataset, by):
+def average_dataset_by_slow(dataset, by):
```
Comment (JasperVanDenBosch) [inline rsatoolbox/data/computations.py]: I think if the slow one is not referenced anywhere you can just remove it. If we need to see how it used to be implemented we can use git history right?

**Review by JasperVanDenBosch — APPROVED** [review]
Looks good to me!

File: rsatoolbox/data/computations.py (hunk 1 above)
Comment (HeikoSchuett) [reply rsatoolbox/data/computations.py]: yes, of course, this will be gone when I merge this in

**Comment by smazurchuk** [comment]
Hey @HeikoSchuett , thanks for looking into this. It does seem that this change sped-up the calculation of RDM's, but in a searchlight type analysis I didn't see any net effect. Linked here is a notebook. I was wondering if you could look at it to see if maybe I did something improperly? I'm also not really sure the best way to import a working branch. 

The imported folder is on the proper branch
image

https://github.com/smazurchuk/general/blob/master/speedup_analysis.ipynb

Thanks!

**Comment by smazurchuk** [comment]
Whoops, realized it is not loading the proper file. Is there a way to load both the package in the python path, and also import the working packing (under the speed branch)  with a separate name?

(actual improvement)
image

**Comment by JasperVanDenBosch** [comment]
> Is there a way to load both the package in the python path, and also import the working packing (under the speed branch) with a separate name?

You could make two separate virtualenv and give them respective names, e.g. `env-rsatoolbox-0.3` and `env-rsatoolbox-speedup` to make it easier to switch, but loading them both in the same runtime would involve renaming the package names I think..

**Comment by smazurchuk** [comment]
Hi @HeikoSchuett . So I started using this branch and doing a searchlight analysis. I've noticed this to be the case before too, but it seems that after about 8-10 cores, adding any more cores doesn't increase the speed. Intuitively, this probably amounts to an I/O limitation. If you log into a node running the job and run the `htop` command, you can see that even though CPU usage looks like it's 100% across the cores, the color is red, indicating a potential I/O limitation

image

There is only run process that is all green, so I think that somewhere there is still some python command (like a list comprehension) that is really limiting any gains in speed. For reference, if I just use `pdist` with correlation coefficients, I usually see all cores used and all green (indicating there is no underlying problem accessing numpy arrays to keep up with 24 cores)

Any thoughts or interest in maybe discussing this further?

Thanks!

**Comment by HeikoSchuett** [comment]
Hi @smazurchuk,
that does sound interesting, as it does indicate some issue with the parallelization. It seems to be somewhat different from the issue I am fixing here though and this seems to help already. I'll thus first merge this into main and raise the parallelization issue you raise as a new issue.

**Pull request merged**